### PR TITLE
New Humanoid Sound script

### PR DESCRIPTION
**Pull request by spotco**
New Humanoid Sound script that's several times more efficient network traffic-wise.
For a networked player in the "worst" case (repeatedly jumping up and down on an upwards slope), reduces network traffic by up to 6x.

TODO: flag loading in cpp code

### CODE REVIEW

File: PlayerScripts/PlayerScriptsPlace/StarterPlayer/StarterCharacterScripts/Sound.rbxmx [hunk 1]
```diff
@@ -0,0 +1,381 @@
...
+	These sounds are used by the "LocalSound" script.
+]]--
+--humanoidSoundNewLocal.rbxmx--
+
+function CreateNewSound(name, id, looped, pitch)
+	local sound = Instance.new("Sound")
+	sound.SoundId = id
+	sound.Name = name
+	sound.archivable = false
+	sound.Parent = script.Parent.Head
```
Comment (hockeychaos) [inline PlayerScripts/PlayerScriptsPlace/StarterPlayer/StarterCharacterScripts/Sound.rbxmx]: is this script created on the server? I it seems be of Script type, not LocalScript. Is this to allow FE games to hear the sounds of humanoids? Also, should we verify Head exists?

File: PlayerScripts/PlayerScriptsPlace/StarterPlayer/StarterCharacterScripts/Sound.rbxmx [hunk 2]
```diff
@@ -0,0 +1,381 @@
+<roblox xmlns:xmime="http://www.w3.org/2005/05/xmlmime" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:noNamespaceSchemaLocation="http://www.roblox.com/roblox.xsd" version="4">
+	<External>null</External>
+	<External>nil</External>
+	<Item class="Script" referent="RBXc83eeca86f73463493b2478b693fc06d">
+		<Properties>
+			<bool name="Disabled">false</bool>
+			<Content name="LinkedSource"><null></null></Content>
+			<string name="Name">Sound</string>
+			<string name="ScriptGuid"></string>
+			<ProtectedString name="Source"><![CDATA[--[[
```
Comment (hockeychaos) [inline PlayerScripts/PlayerScriptsPlace/StarterPlayer/StarterCharacterScripts/Sound.rbxmx]: I'd put yourself as author here

File: PlayerScripts/PlayerScriptsPlace/StarterPlayer/StarterCharacterScripts/Sound.rbxmx [hunk 3]
```diff
@@ -0,0 +1,381 @@
...
+	FreeFalling = 6;
+	FallingDown = 7;
+	Landing = 8;
+	Splash = 9;
+}
+
+local Humanoid = nil
+local Head = nil
+
+--SFX ID to Sound object
```
Comment (hockeychaos) [inline PlayerScripts/PlayerScriptsPlace/StarterPlayer/StarterCharacterScripts/Sound.rbxmx]: why not just use the strings we used in the server script? or have the server use a similar table

File: PlayerScripts/PlayerScriptsPlace/StarterPlayer/StarterCharacterScripts/Sound.rbxmx [hunk 4]
```diff
@@ -0,0 +1,381 @@
...
+	YForLineGivenXAndTwoPts = function(x,pt1x,pt1y,pt2x,pt2y)
+		--(y - y1)/(x - x1) = m
+		local m = (pt1y - pt2y) / (pt1x - pt2x)
+		--float b = pt1.y - m * pt1.x;
+		local b = (pt1y - m * pt1x)
+		return m * x + b
+	end;
+	
+	--Clamps the value of "val" between the "min" and "max"
+	Clamp = function(val,min,max)
```
Comment (hockeychaos) [inline PlayerScripts/PlayerScriptsPlace/StarterPlayer/StarterCharacterScripts/Sound.rbxmx]: we should consider putting this in our math library...

File: PlayerScripts/PlayerScriptsPlace/StarterPlayer/StarterCharacterScripts/Sound.rbxmx [hunk 5]
```diff
@@ -0,0 +1,381 @@
...
+		return math.min(max,math.max(min,val))	
+	end;
+	
+	--Gets the horizontal (x,z) velocity magnitude of the given part
```
Comment (hockeychaos) [inline PlayerScripts/PlayerScriptsPlace/StarterPlayer/StarterCharacterScripts/Sound.rbxmx]: this and the one below I think are technically speed actually, since we remove directionality

File: PlayerScripts/PlayerScriptsPlace/StarterPlayer/StarterCharacterScripts/Sound.rbxmx [hunk 6]
```diff
@@ -0,0 +1,381 @@
...
+	VerticalVelocityMagnitude = function(Head)
+		return math.abs(Head.Velocity.Y)
+	end;
+	
+	--Checks if Sound.Playing is usable
+	SoundIsV2 = function(sound)
+		return sound.IsLoaded
+	end;
+	
+	--Setting Playing/TimePosition values directly result in less network traffic than Play/Pause/Resume/Stop
```
Comment (hockeychaos) [inline PlayerScripts/PlayerScriptsPlace/StarterPlayer/StarterCharacterScripts/Sound.rbxmx]: Is this is the big thing that saves network traffic? does anything else here? If not, why not just modify current script?

File: PlayerScripts/PlayerScriptsPlace/StarterPlayer/StarterCharacterScripts/Sound.rbxmx [hunk 7]
```diff
@@ -0,0 +1,381 @@
...
+
+do
+	-- List of all active Looped sounds
+	local activeLoopedSounds = {}
+	
+	-- Last seen Enum.HumanoidStateType
+	local activeState = nil
+	
+	-- Verify and set that the sound's .Looped property is the value of "looped"
+	function verifyAndSetLoopedForSound(sound, looped)
```
Comment (hockeychaos) [inline PlayerScripts/PlayerScriptsPlace/StarterPlayer/StarterCharacterScripts/Sound.rbxmx]: this seems really wordy, why not just setLooped? Also why is this necessary? Internally it looks like we are already doing this check for you

File: PlayerScripts/PlayerScriptsPlace/StarterPlayer/StarterCharacterScripts/Sound.rbxmx [hunk 8]
```diff
@@ -0,0 +1,381 @@
...
+		if sound.Looped ~= looped then
+			sound.Looped = looped
+		end
+	end
+	
+	-- Verify and set that "sound" is in "activeLoopedSounds".
+	function setSoundInActiveLooped(sound)
```
Comment (hockeychaos) [inline PlayerScripts/PlayerScriptsPlace/StarterPlayer/StarterCharacterScripts/Sound.rbxmx]: why is this not a part of setLooped?

File: PlayerScripts/PlayerScriptsPlace/StarterPlayer/StarterCharacterScripts/Sound.rbxmx [hunk 9]
```diff
@@ -0,0 +1,381 @@
...
+		for i=1, #activeLoopedSounds do
+			if activeLoopedSounds[i] == sound then
+				return
+			end
+		end	
+		table.insert(activeLoopedSounds,sound)
+	end
+	
+	-- Stop all active looped sounds except parameter "except". If "except" is not passed, all looped sounds will be stopped.
+	function stopActiveLoopedSoundsExcept(except)
```
Comment (hockeychaos) [inline PlayerScripts/PlayerScriptsPlace/StarterPlayer/StarterCharacterScripts/Sound.rbxmx]: what does the word active add in this context?

File: PlayerScripts/PlayerScriptsPlace/StarterPlayer/StarterCharacterScripts/Sound.rbxmx [hunk 10]
```diff
@@ -0,0 +1,381 @@
...
+			stateUpdated(Enum.HumanoidStateType.Running)
+		end;
+		
+		[Enum.HumanoidStateType.Running] = function()	
+			local sound = Sounds[SFX.Running]
+			verifyAndSetLoopedForSound(sound,true)
+			stopActiveLoopedSoundsExcept(sound)
+			
+			if Util.HorizontalVelocityMagnitude(Head) > 0.5 then
+				if not sound.IsPlaying then
```
Comment (hockeychaos) [inline PlayerScripts/PlayerScriptsPlace/StarterPlayer/StarterCharacterScripts/Sound.rbxmx]: is this if check really necessary? I don't think it will save much time, and if it does why not just put it in util?

File: PlayerScripts/PlayerScriptsPlace/StarterPlayer/StarterCharacterScripts/Sound.rbxmx [hunk 11]
```diff
@@ -0,0 +1,381 @@
...
+				splashSound.Volume = Util.Clamp(
+					Util.YForLineGivenXAndTwoPts(
+						Util.VerticalVelocityMagnitude(Head), 
+						100, 0.28, 
+						350, 1),
+					0,1)
+				Util.Play(splashSound)
+			end
+			
+			do
```
Comment (hockeychaos) [inline PlayerScripts/PlayerScriptsPlace/StarterPlayer/StarterCharacterScripts/Sound.rbxmx]: is the do here just for scoping or logical separation?

File: PlayerScripts/PlayerScriptsPlace/StarterPlayer/StarterCharacterScripts/Sound.rbxmx [hunk 12]
```diff
@@ -0,0 +1,381 @@
...
+			verifyAndSetLoopedForSound(sound,false)
+			Util.Play(sound)
+		end;
+		
+		[Enum.HumanoidStateType.Freefall] = function()
+			if activeState == Enum.HumanoidStateType.Freefall then
+				return
+			end
+			local sound = Sounds[SFX.FreeFalling]
+			if sound.Volume ~= 0 then
```
Comment (hockeychaos) [inline PlayerScripts/PlayerScriptsPlace/StarterPlayer/StarterCharacterScripts/Sound.rbxmx]: why might as well remove this if

File: PlayerScripts/PlayerScriptsPlace/StarterPlayer/StarterCharacterScripts/Sound.rbxmx [hunk 13]
```diff
@@ -0,0 +1,381 @@
...
+	-- Handle state event fired or OnChange fired
+	function stateUpdated(state)
+		if stateUpdateHandler[state] ~= nil then
+			stateUpdateHandler[state]()
+		end
+		activeState = state
+	end
+	
+	-- Runs on heartbeat
+	function onHeartbeat(step)
```
Comment (hockeychaos) [inline PlayerScripts/PlayerScriptsPlace/StarterPlayer/StarterCharacterScripts/Sound.rbxmx]: can we not do this with events? it seems like it would be more performant to not need a loop here. If a loop is necessary why not spawn or delay a function in the freefall state that does this?

File: PlayerScripts/PlayerScriptsPlace/StarterPlayer/StarterCharacterScripts/Sound.rbxmx (hunk 6 above)
Comment (spotco) [inline PlayerScripts/PlayerScriptsPlace/StarterPlayer/StarterCharacterScripts/Sound.rbxmx]: So it's 4 big savings:
1. Stopping only the active looped sounds (stopActiveLoopedSoundsExcept). Old script would stop all looped sounds (always stop swimming,running and climbing), even if they weren't playing.

2. Use of .Playing and .TimePosition if they're available (instead of Play/Pause/Resume/Stop). The Play methods are more network traffic because they've got additional replication things they need to do (update the internal "lastSetLuaTime", fire Played/Paused/Resumed/Stopped events)

3. Changes to the behaviour of the "Running" sound and "Falling" sound to not trigger quite as often. These are subtle behaviour changes, but what the player hears ingame should be indistinguishable.

4. Avoids any normally un-noticeable changes to Volume/Pitch (there were a couple of not-very-functional speed-to-pitch linear relationships in the old code, they're just set-once pitches now). This is only network traffic in non-filtering enabled games.

File: PlayerScripts/PlayerScriptsPlace/StarterPlayer/StarterCharacterScripts/Sound.rbxmx (hunk 7 above)
Comment (spotco) [inline PlayerScripts/PlayerScriptsPlace/StarterPlayer/StarterCharacterScripts/Sound.rbxmx]: youre right this will be changed

File: PlayerScripts/PlayerScriptsPlace/StarterPlayer/StarterCharacterScripts/Sound.rbxmx (hunk 8 above)
Comment (spotco) [inline PlayerScripts/PlayerScriptsPlace/StarterPlayer/StarterCharacterScripts/Sound.rbxmx]: this is a little different:
so this localscript maintains a list of "looped sounds" that are actively playing.
this "setSoundInActiveLooped" adds the parameter sound to the "activeLoopedSounds" array.

File: PlayerScripts/PlayerScriptsPlace/StarterPlayer/StarterCharacterScripts/Sound.rbxmx (hunk 9 above)
Comment (spotco) [inline PlayerScripts/PlayerScriptsPlace/StarterPlayer/StarterCharacterScripts/Sound.rbxmx]: this modifies the "activeLoopedSounds" array

File: PlayerScripts/PlayerScriptsPlace/StarterPlayer/StarterCharacterScripts/Sound.rbxmx (hunk 10 above)
Comment (spotco) [inline PlayerScripts/PlayerScriptsPlace/StarterPlayer/StarterCharacterScripts/Sound.rbxmx]: yeah this one can go into Util

File: PlayerScripts/PlayerScriptsPlace/StarterPlayer/StarterCharacterScripts/Sound.rbxmx (hunk 11 above)
Comment (spotco) [inline PlayerScripts/PlayerScriptsPlace/StarterPlayer/StarterCharacterScripts/Sound.rbxmx]: logical separation

File: PlayerScripts/PlayerScriptsPlace/StarterPlayer/StarterCharacterScripts/Sound.rbxmx (hunk 12 above)
Comment (spotco) [inline PlayerScripts/PlayerScriptsPlace/StarterPlayer/StarterCharacterScripts/Sound.rbxmx]: yes

File: PlayerScripts/PlayerScriptsPlace/StarterPlayer/StarterCharacterScripts/Sound.rbxmx (hunk 2 above)
Comment (spotco) [inline PlayerScripts/PlayerScriptsPlace/StarterPlayer/StarterCharacterScripts/Sound.rbxmx]: ok

File: PlayerScripts/PlayerScriptsPlace/StarterPlayer/StarterCharacterScripts/Sound.rbxmx (hunk 3 above)
Comment (spotco) [inline PlayerScripts/PlayerScriptsPlace/StarterPlayer/StarterCharacterScripts/Sound.rbxmx]: ill use object.Name

File: PlayerScripts/PlayerScriptsPlace/StarterPlayer/StarterCharacterScripts/Sound.rbxmx (hunk 4 above)
Comment (spotco) [inline PlayerScripts/PlayerScriptsPlace/StarterPlayer/StarterCharacterScripts/Sound.rbxmx]: where is this

File: PlayerScripts/PlayerScriptsPlace/StarterPlayer/StarterCharacterScripts/Sound.rbxmx (hunk 5 above)
Comment (spotco) [inline PlayerScripts/PlayerScriptsPlace/StarterPlayer/StarterCharacterScripts/Sound.rbxmx]: name will b changed

File: PlayerScripts/PlayerScriptsPlace/StarterPlayer/StarterCharacterScripts/Sound.rbxmx (hunk 1 above)
Comment (spotco) [inline PlayerScripts/PlayerScriptsPlace/StarterPlayer/StarterCharacterScripts/Sound.rbxmx]: old code doesnt verify that but it can be done

File: PlayerScripts/PlayerScriptsPlace/StarterPlayer/StarterCharacterScripts/Sound.rbxmx (hunk 4 above)
Comment (hockeychaos) [inline PlayerScripts/PlayerScriptsPlace/StarterPlayer/StarterCharacterScripts/Sound.rbxmx]: I just mean the standard lua math lib

File: PlayerScripts/PlayerScriptsPlace/StarterPlayer/StarterCharacterScripts/Sound.rbxmx (hunk 8 above)
Comment (hockeychaos) [inline PlayerScripts/PlayerScriptsPlace/StarterPlayer/StarterCharacterScripts/Sound.rbxmx]: yeah it just seems like if we have a function called setLooped we should also call setSoundInActiveLooped if looped == true. Or just remove this function entirely and have it as part of setLooped.

File: PlayerScripts/PlayerScriptsPlace/StarterPlayer/StarterCharacterScripts/Sound.rbxmx (hunk 9 above)
Comment (hockeychaos) [inline PlayerScripts/PlayerScriptsPlace/StarterPlayer/StarterCharacterScripts/Sound.rbxmx]: the word active is a bit vague to me. Does that mean it is playing? Does it mean it can play? Does it mean it is the main sound that is playing?

Comment (spotco) [inline PlayerScripts/PlayerScriptsPlace/StarterPlayer/StarterCharacterScripts/Sound.rbxmx]: ill rename active -> playing.
It's "an" active looped sound that is playing. For any state they may be one, multiple or none. Depends on the state handler implementation.

File: PlayerScripts/PlayerScriptsPlace/StarterPlayer/StarterCharacterScripts/Sound.rbxmx (hunk 13 above)
Comment (spotco) [inline PlayerScripts/PlayerScriptsPlace/StarterPlayer/StarterCharacterScripts/Sound.rbxmx]: Heartbeat's here for some new behaviour:
1. Only trigger the freefalling sound if the last active state was "FreeFall", and vertical speed is above 75 (arbitrary number that made it NOT trigger for normal jumps). 

2. If last active state was running, and horizontal speed is less than 0.5, stop the running sound playback.

I personally think it's a lot cleaner to do this with a stateless update method (or else we'd need to manage/preempt the spawn or delay functions). The only "state" that this code has right now is "lastActiveState" and "activeLoopedSounds" (soon to be renamed playingLoopedSounds).

And they don't seem to be a big problem performance-wise. Heartbeat stuff generally shouldn't be, is that correct? (It's one heartbeat listener per client too)

File: PlayerScripts/PlayerScriptsPlace/StarterPlayer/StarterCharacterScripts/Sound.rbxmx (hunk 3 above)
Comment (spotco) [inline PlayerScripts/PlayerScriptsPlace/StarterPlayer/StarterCharacterScripts/Sound.rbxmx]: Thinking about it:
This is to keep the SFX.FreeFalling syntax (as opposed to SFX["FreeFalling"]).
The first would be better in a typed language because of compiler checking, I'm guessing the errors in lua would be about the same.
Anyway, I'm now for keeping it as is (SFX.FallingDown) because easier to read?

File: PlayerScripts/PlayerScriptsPlace/StarterPlayer/StarterCharacterScripts/Sound.rbxmx (hunk 4 above)
Comment (spotco) [inline PlayerScripts/PlayerScriptsPlace/StarterPlayer/StarterCharacterScripts/Sound.rbxmx]: re: I'll do another PR sometime with some more math_whatever functions.
I see lua_pushnumber and immediately go nope.

File: PlayerScripts/PlayerScriptsPlace/StarterPlayer/StarterCharacterScripts/Sound.rbxmx (hunk 13 above)
Comment (hockeychaos) [inline PlayerScripts/PlayerScriptsPlace/StarterPlayer/StarterCharacterScripts/Sound.rbxmx]: also remember that this has to run on a mobile device in VR, so everything counts.

File: PlayerScripts/PlayerScriptsPlace/StarterPlayer/StarterCharacterScripts/Sound.rbxmx [hunk 14]
```diff
@@ -0,0 +1,382 @@
...
+		return math.min(max,math.max(min,val))	
+	end;
+	
+	--Gets the horizontal (x,z) velocity magnitude of the given part
+	HorizontalSpeed = function(Head)
+		local hVel = Head.Velocity + Vector3.new(0,-Head.Velocity.Y,0)
+		return hVel.magnitude	
+	end;
+	
+	TickSpeedSecondsAndLengthSecondsToNormalizedIncrement = function(tickSpeedSeconds, lengthSeconds)
```
Comment (hockeychaos) [inline PlayerScripts/PlayerScriptsPlace/StarterPlayer/StarterCharacterScripts/Sound.rbxmx]: Is this function necessary? The name is kind of confusing and it is longer than the actual function

Comment (spotco) [inline PlayerScripts/PlayerScriptsPlace/StarterPlayer/StarterCharacterScripts/Sound.rbxmx]: u got a problem with longer names m8?

(Anyway, this math is necessary to keep the update timestep variable, and the unit of the number in the application code in seconds)

Comment (hockeychaos) [inline PlayerScripts/PlayerScriptsPlace/StarterPlayer/StarterCharacterScripts/Sound.rbxmx]: oh I just meant lets just put the code its doing directly into the function, since its just a divide and its the only place we use it.
